File: quick_sort.c
#include <emscripten/emscripten.h>

EMSCRIPTEN_KEEPALIVE
void quick_sort(int* arr, int left, int right) {
    int i = left, j = right;
    int pivot = arr[(left + right) / 2];
    while (i <= j) {
        while (arr[i] < pivot) i++;
        while (arr[j] > pivot) j--;
        if (i <= j) {
            int tmp = arr[i];
            arr[i] = arr[j];
            arr[j] = tmp;
            i++; j--;
        }
    }
    if (left < j) quick_sort(arr, left, j);
    if (i < right) quick_sort(arr, i, right);
}

File: quick_sort.js
function quickSortJS(arr, left=0, right=arr.length-1){
    if(left >= right) return;
    let pivot = arr[Math.floor((left+right)/2)];
    let i = left, j = right;
    while(i <= j){
        while(arr[i] < pivot) i++;
        while(arr[j] > pivot) j--;
        if(i <= j){
            [arr[i], arr[j]] = [arr[j], arr[i]];
            i++; j--;
        }
    }
    if(left < j) quickSortJS(arr, left, j);
    if(i < right) quickSortJS(arr, i, right);
}
